### PR TITLE
Change workspace with mouse wheel

### PR DESCRIPTION
**Pull request by deklov**

### CODE REVIEW

File: swaybar/bar.c [hunk 1]
```diff
@@ -89,6 +89,13 @@ static void mouse_button_notify(struct window *window, int x, int y, uint32_t bu
 	}
 }
 
+static void mouse_scroll_notify(struct window *window, enum scroll_direction direction) {
+	sway_log(L_DEBUG, "Mouse wheel scrolled %s", direction == SCROLL_UP ? "up" : "down");
+
+	const char *workspace_name = direction == SCROLL_UP ? "prev" : "next";
```
Comment (deklov) [inline swaybar/bar.c]: Not sure what the correct behavior is here: prev vs. prev_on_output.

File: include/client/window.h [hunk 2]
```diff
@@ -27,11 +27,18 @@ struct cursor {
...
+	SCROLL_DOWN
+};
+
 struct pointer_input {
 	int last_x;
 	int last_y;
 
 	void (*notify_button)(struct window *window, int x, int y, uint32_t button);
+
+	void (*notify_scroll)(struct window *window, enum scroll_direction direction);
```
Comment (deklov) [inline include/client/window.h]: I guess we could pass time as well but I didn't needed it..

File: wayland/window.c [hunk 3]
```diff
@@ -47,7 +47,17 @@ static void pointer_handle_button(void *data, struct wl_pointer *pointer, uint32
 }
 
 static void pointer_handle_axis(void *data, struct wl_pointer *pointer,
-		    uint32_t time, uint32_t axis, wl_fixed_t value) {
+				uint32_t time, uint32_t axis, wl_fixed_t value) {
```
Comment (deklov) [inline wayland/window.c]: Not exactly sure what axis and value. Documentation is lacking. Testing suggests that on a mouse with a standard scroll wheel axis is zero and value is -10.0/10.0 when scrolled up/down.

**Comment by deklov** [comment]
This is more of a  RFC pull request.

File: swaybar/bar.c (hunk 1 above)
Comment (deklov) [inline swaybar/bar.c]: Btw. One annoying thing with prev/next is that when changing to a workspace on a different output the mouse pointer follows.

**Comment by Hummer12007** [comment]
@deklov there's horizontal scrolling too (axis == 0 -> vertical, 1 -> horizontal)

File: swaybar/bar.c (hunk 1 above)
Comment (SirCmpwn) [inline swaybar/bar.c]: The correct behavior is prev_on_output.

File: wayland/window.c [hunk 4]
```diff
@@ -41,13 +41,23 @@ static void pointer_handle_button(void *data, struct wl_pointer *pointer, uint32
...
+		window->pointer_input.notify_button(window, input->last_x, input->last_y, button);
 	}
 }
 
 static void pointer_handle_axis(void *data, struct wl_pointer *pointer,
-		    uint32_t time, uint32_t axis, wl_fixed_t value) {
+				uint32_t time, uint32_t axis, wl_fixed_t value) {
+	struct window *window = data;
+
+	if (!sway_assert(!axis, "Only support horizontal axis")) {
```
Comment (SirCmpwn) [inline wayland/window.c]: Assertions are to detect things that should never happen. Horizontal scrolling doesn't qualify.

**Comment by SirCmpwn** [comment]
LGTM

File: wayland/window.c [hunk 5]
```diff
@@ -47,7 +47,26 @@ static void pointer_handle_button(void *data, struct wl_pointer *pointer, uint32
...
+		direction = wl_fixed_to_double(value) < 0 ? SCROLL_UP : SCROLL_DOWN;
+		break;
+	case 1:
+		direction = wl_fixed_to_double(value) < 0 ? SCROLL_LEFT : SCROLL_RIGHT;
+		break;
+	default:
+		if (!sway_assert(false, "Only support horizontal axis")) {
+			return;
+		}
+	}
```
Comment (deklov) [inline wayland/window.c]: @Hummer12007 something like this?

Comment (Hummer12007) [inline wayland/window.c]: Yeah, but why should you add a default clause? The axis enum in libinput/wayland has only two directions.

Comment (SirCmpwn) [inline wayland/window.c]: I don't have a strong preference for or against a default here, but the log message needs to be updated.

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**